calculus.tex

```diff
@@ -0,0 +1,81 @@
+
+\paragraph{Rules}
+
+Below, we give some rules for computing derivatives.
+Unless otherwise indicated, $x$ has shape $\mathcal{S}$, and $u$ and $v$ are dependent on $x$ and have shapes $\mathcal{U}$ and $\mathcal{V}$, respectively.
+Define $I_\mathcal{S}$, a generalization of the identity matrix:
+\begin{align*}
+  I_\mathcal{S} &\in \reals^{\mathcal{S} \times \mathcal{S'}} \\
+  [I_\mathcal{S}]_{s, s'} &= \begin{cases}
+    1 & \text{if $s = s'$} \\
+    0 & \text{otherwise.}
+  \end{cases}
+\end{align*}
+
+\begin{align*}
+  \ddx x &= I_\mathcal{S} \\
+  \ddx u &= 0 && \text{$u$ does not depend on $x$} \\
+  \ddx f(u) &= f'(u) \odot \ddx u && f \colon \reals \rightarrow \reals \\
+  \ddx (u + v) &= \ddx u + \ddx v \\
+  \ddx \nsum{ax} u &= \nsum{ax} \ddx u \\
+  \ddx (u \odot v) &= \ddx u \odot v + u \odot \ddx v \\
+  \ddx (u \ndot{ax} v) &= \ddx u \ndot{ax} v + u \ndot{ax} \ddx v \\
+  \ddx \frac{u}{v} &= \frac{\ddx u \odot v - u \odot \ddx v}{v^2} \\
+  \ddx u_{\nmov{ax1}{ax2}} &= \left[\ddx u\right]_{\nmov{ax1}{ax2}}
+\end{align*}
+
+Here's an example using these rules to derive a new one:
+\begin{align*}
+  \ddx (\nfun{ax}{softmax} u) &= \ddx \frac{\exp u}{\nsum{ax} \exp u} \\
+    &= \frac{\exp u \odot \ddx u \odot \nsum{ax} \exp u - \exp u \odot \nsum{ax} (\exp u \odot \ddx u)}{(\nsum{ax} \exp u)^2} \\
+    &= \nfun{ax}{softmax} u \odot \left(\ddx u - \nfun{ax}{softmax} u \ndot{ax} \ddx u\right).
+\end{align*}
+
+\paragraph{Chain rule}
+
+Above we gave a chain rule for elementwise operations. The general chain rule looks like this for functions of one and two variables; three or more variables are analogous.
+\begin{align*}
+  \ddx f(u) &= \frac{\partial}{\partial u} f(u) \ndot{\mathcal{U}' \mid \mathcal{U}} \ddx u \\
+  \ddx f(u, v) &= \frac{\partial}{\partial u} f(u, v) \ndot{\mathcal{U}' \mid \mathcal{U}} \ddx u + \frac{\partial}{\partial v} f(u, v) \ndot{\mathcal{V}' \mid \mathcal{V}} \ddx v
+\end{align*}
+where $\ndot{ax1|ax2}$ contracts $\name{ax1}$ in the left operand with $\name{ax2}$ in the right operand: $A \ndot{ax1|ax2} B = \sum_i A_{\nidx{ax1}{i}} \odot B_{\nidx{ax2}{i}}$.
+
+\paragraph{Broadcasting}
+
+If $f \colon \reals^\mathcal{S} \rightarrow \reals^\mathcal{T}$, then recall that $f$ can be extended to $\reals^{\mathcal{S} \cup \mathcal{S^+}}$ where $\mathcal{S}$ and $\mathcal{S^+}$ are orthogonal.
+
+It's more convenient here to notate the derivative of $f$ as $Df$. If $f$ has two arguments, its partial derivatives are $D_1 f$ and $D_2 f$.
+
+Although $Df$ extends to $\reals^{\mathcal{S} \cup \mathcal{S^+}}$ using the usual broadcasting rules, the extension of the derivative is unfortunately not the derivative of the extension. To avoid confusion, write $f^+$ for the extension:
+\begin{align*}
+  f^+ \colon \reals^{\mathcal{S} \cup \mathcal{S^+}} &\rightarrow \reals^{\mathcal{T} \cup \mathcal{S^+}} \\
+  f^+(x)_r &= f(x_r).
+\end{align*}
+Then the derivative of $f^+$ is:
```

Comment:
This is pretty interesting. I guess this is the negative side of not having an explicit "extend / vmap" function that can be differentiated. 

I think this is relatively intuitive though?